dev/src/bundle.ts

```diff
@@ -0,0 +1,184 @@
+    } else if (querySnapshot instanceof QuerySnapshot) {
+      this.addNamedQuery(documentOrName, querySnapshot);
+    }
+
+    return this;
+  }
+
+  private addBundledDocument(snap: DocumentSnapshot) {
+    if (snap.exists) {
+      const docProto = snap.toDocumentProto();
+      if (
+        !this.documents.has(snap.id) ||
+        Timestamp.fromProto(this.documents.get(snap.id)!.metadata.readTime!) <
+          snap.readTime
+      ) {
+        this.documents.set(snap.id, {
+          document: docProto,
+          metadata: {
+            documentKey: docProto.name,
+            readTime: snap.readTime.toProto().timestampValue,
+          },
+        });
+      }
+      if (snap.readTime > this.latestReadTime) {
+        this.latestReadTime = snap.readTime;
+      }
+    }
+  }
+
+  private addNamedQuery(name: string, querySnap: QuerySnapshot) {
+    if (this.namedQueries.has(name)) {
+      throw new Error(`Query name conflict: ${name} is already added.`);
+    }
+
+    this.namedQueries.set(name, {
+      name,
+      bundledQuery: querySnap.query.toBundledQuery(),
+      readTime: querySnap.readTime.toProto().timestampValue,
+    });
+
+    for (const snap of querySnap.docs) {
+      this.addBundledDocument(snap);
+    }
+
+    if (querySnap.readTime > this.latestReadTime) {
+      this.latestReadTime = querySnap.readTime;
+    }
+  }
+
+  private lengthPrefixedBuffer(payload: string): Buffer {
+    const buffer = Buffer.from(payload, 'utf-8');
+    const lengthBuffer = Buffer.from(buffer.length.toString());
+    return Buffer.concat([lengthBuffer, buffer]);
+  }
+
+  stream(): NodeJS.ReadableStream {
+    const readable = new Readable({
+      objectMode: false,
+      read: () => {
+        const metadata: firestore.IBundleMetadata = {
+          id: this.bundleId,
+          createTime: this.latestReadTime.toProto().timestampValue,
+        };
+        readable.push(
+          this.lengthPrefixedBuffer(
+            JSON.stringify({metadata} as firestore.IBundleElement)
+          )
+        );
+
+        for (const namedQuery of this.namedQueries.values()) {
+          readable.push(
+            this.lengthPrefixedBuffer(
+              JSON.stringify({namedQuery} as firestore.IBundleElement)
+            )
+          );
+        }
+
+        for (const bundledDocument of this.documents.values()) {
+          const documentMetadata: firestore.IBundledDocumentMetadata =
+            bundledDocument.metadata;
+          const document: api.IDocument = bundledDocument.document;
+
+          readable.push(
+            this.lengthPrefixedBuffer(
+              JSON.stringify({documentMetadata} as firestore.IBundleElement)
+            )
+          );
+          readable.push(
+            this.lengthPrefixedBuffer(
+              JSON.stringify({document} as firestore.IBundleElement)
+            )
+          );
+        }
+        readable.push(null);
+      },
+    });
+
+    return readable;
```

Comment:
You are building the stream contents before returning them, which takes away some of the advantage of using a stream to begin with. Take a look at https://nodesource.com/blog/understanding-streams-in-nodejs/ (search for "async iterator").

Otherwise, I think you might need to move your `push` calls outside of `read()`.

Reply:
There is no code in here that enforces that enforces that no documents are added after Stream is called. We should either add this or support the following use pattern:

```
const stream = bundle.stream();
bundle.add()
bundle.add()
```

Since we can support this API usage with no additional complexity, I think we should evaluate offering it. We can even directly instantiate a stream when we create a BundleBuilder and push directly to it instead of storing the documents in temporary variables. While we still need to keep the read times, we don't need to keep the contents once the user has read the data which reduces our memory consumption.

If you want rather want to add enforcement that all data has been added before invoking `stream`, I would still like to encourage you to re-arrange the code here to push data outside of the `read` call. You could also simply use a PassThrough stream: 

```
const readable = new PassThrough();
readable.push('...');
readable.push('...');
readable.push('...');
return readable;
```